### DISCUSSION TITLE
empty strings and `'failed` indicator in `consult--dynamic-compute`

### DISCUSSION THREAD

**Discussion by ahmed-shariff** [minad/consult#1117] [Q&A]
Hi, 

I have been hacking together a way to narrow searching org-roam nodes using [org-roam-ql](https://github.com/ahmed-shariff/org-roam-ql). The idea was to have org-roam-ql process the async part to filter nodes (e.g., nodes with TODO or INPROGRESS). I have gotten most of it working. I was hoping I could get help with two pieces:
1. I want all nodes to be shown when there is an error or the async string is empty. consult handles empty strings differently. As a workaround, I intercept the split call and return "-" as the async-string when the split function returns an empty string. Is there a better way to handle this situation?
2. When org-roam-ql fails to process the async string, I want to show the "failed" indicator. However, the sink is wrapped with the indicator functions inside `consult--dynamic-compute`. Do you have any suggestions on how I might handle that?

The implementation I have for now:
```emacs-lisp
(defun consult-org-roam-ql ()
    "Consult with org-roam-ql for searching/narrowing."
    (interactive)
    (minibuffer-with-setup-hook
        (lambda ()
          ;; KLUDGE: No idea why this is here!
          (set-syntax-table emacs-lisp-mode-syntax-table)
          (add-hook 'completion-at-point-functions
                    #'org-roam-ql--completion-at-point nil t))
    (let* ((style (consult--async-split-style))
           (fn (plist-get style :function))
           ;; Override how the empty string is handled!
           ;; When empty async-str should return default candidates
           (split (lambda (str)
                    (pcase-let* ((res (funcall fn str style))
                                 (`(,async-str ,_ ,force . ,_) res))
                      (when (and force (equal "" async-str))
                        (setf (car res) "-"))
                      res)))
           ;; Default candidates
           (nodes (mapcar (lambda (node)
                            (cons (org-roam-node-title node) node))
                          (org-roam-node-list)))
           ;; The sink is what holds the candidates and feed it back to all-completions
           (sink (consult--async-sink)))

      ;; Feeding initial set of candidates to sink
      (funcall sink nodes)
      (-->
       (consult--dynamic-compute
        sink
        (lambda (input) 
          (if (and (> (length input) 0) (not (equal input "-")))
              ;; TODO: can I update the state/indicator somehow?
              (condition-case err
                  (mapcar
                   (lambda (node) (cons (org-roam-node-title node) node))
                   (org-roam-ql-nodes (read input)))
                (user-error nodes))
            nodes)))
       (consult--async-throttle it)
       (consult--async-split it split)
       (consult--read it :initial ";" :keymap org-roam-ql--read-query-map)
       (or (cdr (assoc it nodes))
           (org-roam-node-create :title it))))))
```

**Comment by minad** [comment]
I've recently added support for empty strings to `consult--dynamic-compute` if you set `consult-async-min-input` to 0 or pass 0 to `consult--dynamic-compute` as value for the `MIN-INPUT` argument. This should at least help with your first question.

**Reply by ahmed-shariff** [reply]
Thank you for pointing that out. I'll play around with it and see how it works. cheers!
